### ISSUE TITLE
Regex performance regression on PCRE2

### ISSUE THREAD

**Issue by Blacksmoke16** [crystal-lang/crystal#13144]
## Bug Report

There seems to be a severe performance regression when leveraging PCRE2 over PCRE. For example, given the sample code:

```cr
require "benchmark"

STR = "slug1"

Benchmark.ips do |x|
  x.report("starts_with?") do
    STR.starts_with? /\d/
  end

  x.report("matches?") do
    STR.matches? /^\d/
  end
end
```

```sh
Crystal 1.8.0-dev [6e9c35656] (2023-03-02)
LLVM: 14.0.6
Default target: x86_64-pc-linux-gnu
```

When using PCRE2:
```sh
./bin/crystal run --release -Duse_pcre2 test.cr
starts_with? 394.72k (  2.53µs) (± 3.98%)  20.1kB/op  19.63× slower
    matches?   7.75M (129.07ns) (± 1.71%)    128B/op        fastest
```

While if you go back to PCRE:
```sh
./bin/crystal run --release -Duse_pcre test.cr
starts_with?  19.24M ( 51.98ns) (± 2.84%)  16.0B/op   1.39× slower
    matches?  26.83M ( 37.27ns) (± 2.94%)   0.0B/op        fastest
```

you get a ~50x reduction in performance, and a ~1250x increase in memory in the `starts_with?` case.

I investigated a bit and the root issue seems to be that we're creating a new `match_context` and `match_data` instance on _every_ operation:

https://github.com/crystal-lang/crystal/blob/6e9c35656b5ae9ac89790d99146f7f10d8f21c00/src/regex/pcre2.cr#L157-L158

**Comment by straight-shoota** [comment]
I'm not sure if allocation of match data and context are really the critical difference. Technically, in PCRE the same allocation for match data happens, it's just in the bindings, not the library, and called `ovector`.
And I don't see how we could avoid allocating a new match data for every operation because a pointer to part of the match data is retained in `MatchData` and there's really no way to know whether that is still used before executing the next match.

The match context can probably be shared per thread. Let's try that:

```diff
--- i/src/regex/pcre2.cr
+++ w/src/regex/pcre2.cr
@@ -145,18 +145,20 @@ module Regex::PCRE2
   #
   # Only a single `match` function can run per thread at any given time, so there
   # can't be any concurrent access to the JIT stack.
-  @@jit_stack = Crystal::ThreadLocalValue(LibPCRE2::JITStack).new
+  @@match_context = Crystal::ThreadLocalValue(LibPCRE2::MatchContext).new

-  def self.jit_stack
-    @@jit_stack.get do
-      LibPCRE2.jit_stack_create(32_768, 1_048_576, general_context) || raise "Error allocating JIT stack"
+  def self.match_context
+    @@match_context.get do
+      LibPCRE2.match_context_create(nil).tap do |match_context|
+        jit_stack = LibPCRE2.jit_stack_create(32_768, 1_048_576, general_context) || raise "Error allocating JIT stack"
+        LibPCRE2.jit_stack_assign(match_context, nil, jit_stack.as(Void*))
+      end
     end
   end

   private def match_data(str, byte_index, options)
     match_data = LibPCRE2.match_data_create_from_pattern(@re, Regex::PCRE2.general_context)
-    match_context = LibPCRE2.match_context_create(nil)
-    LibPCRE2.jit_stack_assign(match_context, nil, Regex::PCRE2.jit_stack.as(Void*)) if @jit
+    match_context = @jit ? Regex::PCRE2.match_context : Pointer(Void).null.as(LibPCRE2::MatchContext)
     match_count = LibPCRE2.match(@re, str, str.bytesize, byte_index, pcre2_options(options) | LibPCRE2::NO_UTF_CHECK, match_data, match_context)

     if match_count < 0
```

That brings a considerable improvement for `matches?`, but `starts_with?` is still orders of magnitude out:
 
```console
$ # without patch:
$ bin/crystal run --release -Duse_pcre2 .test/bm-pcre2.cr
starts_with? 277.07k (  3.61µs) (±10.67%)  20.1kB/op   3.46× slower
    matches? 957.82k (  1.04µs) (±44.35%)    128B/op        fastest
$ # with patch:
$ bin/crystal run --release -Duse_pcre2 .test/bm-pcre2.cr
starts_with? 373.75k (  2.68µs) (± 6.45%)  20.1kB/op  40.51× slower
    matches?  15.14M ( 66.05ns) (± 3.99%)    128B/op        fastest
```

**Comment by straight-shoota** [comment]
So I got a better approach. Similar performance characteristics, but it seems to be working correctly (by evidence of the spec suite).
This makes the match context global (like the general context) and the JIT stack 

```diff
diff --git i/src/regex/lib_pcre2.cr w/src/regex/lib_pcre2.cr
index b81de3399..1f2a855da 100644
--- i/src/regex/lib_pcre2.cr
+++ w/src/regex/lib_pcre2.cr
@@ -194,7 +194,7 @@ lib LibPCRE2
   type JITStack = Void*

   fun jit_stack_create = pcre2_jit_stack_create_8(startsize : LibC::SizeT, maxsize : LibC::SizeT, gcontext : GeneralContext) : JITStack
-  fun jit_stack_assign = pcre2_jit_stack_assign_8(mcontext : MatchContext, callable_function : Void*, callable_data : Void*) : Void
+  fun jit_stack_assign = pcre2_jit_stack_assign_8(mcontext : MatchContext, callable_function : Void* -> JITStack, callable_data : Void*) : Void

   fun pattern_info = pcre2_pattern_info_8(code : Code*, what : UInt32, where : Void*) : Int

diff --git i/src/regex/pcre2.cr w/src/regex/pcre2.cr
index 98ba7d9b7..c4209cd28 100644
--- i/src/regex/pcre2.cr
+++ w/src/regex/pcre2.cr
@@ -141,6 +141,10 @@ module Regex::PCRE2
     LibPCRE2.general_context_create(->(size, _data) { GC.malloc(size) }, ->(pointer, _data) { GC.free(pointer) }, nil)
   end

+  class_getter match_context do
+    LibPCRE2.match_context_create(nil)
+  end
+
   # Returns a JIT stack that's shared in the current thread.
   #
   # Only a single `match` function can run per thread at any given time, so there
@@ -155,8 +159,8 @@ module Regex::PCRE2

   private def match_data(str, byte_index, options)
     match_data = LibPCRE2.match_data_create_from_pattern(@re, Regex::PCRE2.general_context)
-    match_context = LibPCRE2.match_context_create(nil)
-    LibPCRE2.jit_stack_assign(match_context, nil, Regex::PCRE2.jit_stack.as(Void*)) if @jit
+    match_context = PCRE2.match_context
+    LibPCRE2.jit_stack_assign(match_context, ->(_data) { Regex::PCRE2.jit_stack }, nil) if @jit
     match_count = LibPCRE2.match(@re, str, str.bytesize, byte_index, pcre2_options(options) | LibPCRE2::NO_UTF_CHECK, match_data, match_context)

     if match_count < 0
```

```console
$ bin/crystal run --release -Duse_pcre2 .test/bm-pcre2.cr
starts_with? 291.96k (  3.43µs) (± 2.92%)  20.1kB/op  50.05× slower
    matches?  14.61M ( 68.43ns) (± 4.55%)    128B/op        fastest
```

Performance of `matches?` should acceptable, though there might still be room for improvement.
But `starts_with?` is really perplexing. I suppose the pretty significant memory allocation might be the cause. But I don't know why it allocates 20kB. It can't be the JIT stack - that would be bigger and would only allocate once per thread.

**Comment by straight-shoota** [comment]
When disabling JIT entirely `matches?` has similar characteristics as `starts_with?`. So it would appear that `starts_with?` doesn't use JIT (probably because of the `ANCHORED` flag?).

The 20kb allocation is a vector that `pcre2_match` allocates for backtracking. This could be avoided if we re-used the match data because this heap would only be allocated once. https://www.pcre.org/current/doc/html/pcre2perform.html#SEC3

So a possible solution for this problem could be to actually store match data in the `Regex` instance and re-use it for subsequent matches. Then we'd just have to dup the ovector to pass to `MatchData`.

**Comment by asterite** [comment]
What's strange is that the memory we report is from the GC... so how come those 20kb from pcre appear on our side? Are we allocating those 20kb somewhere on our side?

**Comment by straight-shoota** [comment]
@asterite PCRE2 is configured to allocate via the GC. This was necessary so that the GC could clean up the pointers in `MatchData`.

**Comment by straight-shoota** [comment]
I think I have this figured out Will post a PR shortly.

**Comment by HertzDevil** [comment]
The docs specifically mention that the anchor flags are unsupported: https://www.pcre.org/current/doc/html/pcre2jit.html#SEC5

**Comment by asterite** [comment]
> The docs specifically mention that the anchor flags are unsupported

Then my suggestion is to check the current code, because we are passing it for `starts_with?`

**Comment by straight-shoota** [comment]
For the scope of this issue, that's fine. The only result of anchor flag being unsupported for JIT is that the matcher doesn't use JIT. But the interpreted approach should be good enough (if implemented correctly, which #13146 should achieve). Anchored expressions don't benefit as much from JIT optimization anyway because one of the main advantages of JIT comes from repeated execution at different start indices.
